File: frontend/app/stores/habits.ts
import { defineStore } from "pinia";
import type { EntryChange, Habit, HabitForm, HabitOverviewItem } from "~~/shared/types/habits";
import { apiFetch } from "~/services/api/client";

export type SortMode = "manual" | "name" | "color" | "score" | "status";

export const useHabitsStore = defineStore("habits", {
  state: () => ({
    items: [] as HabitOverviewItem[],
    loading: false,
    from: "",
    to: "",
    sort: "manual" as SortMode,
    showArchived: false,
  }),
  actions: {
    async loadOverview(from: string, to: string) {
      this.from = from;
      this.to = to;
      this.loading = this.items.length === 0;
      try {
        this.items = await apiFetch<HabitOverviewItem[]>("/habits/overview", {
          query: { from, to, sort: this.sort, include_archived: this.showArchived },
        });
      } finally {
        this.loading = false;
      }
    },

    async reload() {
      if (this.from && this.to) await this.loadOverview(this.from, this.to);
    },

    _applyChange(habitId: number, change: EntryChange) {
      const item = this.items.find((i) => i.habit.id === habitId);
      if (!item) return;
      item.score = change.score;
      item.streak = change.streak;
      const window: Record<string, number> = {};
      for (const [date, value] of Object.entries(change.entries)) {
        if (date >= this.from && date <= this.to) window[date] = value;
      }
      item.entries = window;
    },

    async toggle(habitId: number, date: string) {
      const item = this.items.find((i) => i.habit.id === habitId);
      const previous = item?.entries[date];
      if (item) item.entries[date] = previous === 2 ? 0 : 2; // optimistic guess
      try {
        const change = await apiFetch<EntryChange>(`/habits/${habitId}/entries/${date}/toggle`, {
          method: "POST",
        });
        this._applyChange(habitId, change);
      } catch (error) {
        if (item) {
          if (previous === undefined) delete item.entries[date];
          else item.entries[date] = previous;
        }
        throw error;
      }
    },

    async setValue(habitId: number, date: string, value: number, notes?: string) {
      const change = await apiFetch<EntryChange>(`/habits/${habitId}/entries/${date}`, {
        method: "PUT",
        body: { value, notes },
      });
      this._applyChange(habitId, change);
      const item = this.items.find((i) => i.habit.id === habitId);
      if (item && notes !== undefined) {
        if (notes) item.notes[date] = notes;
        else delete item.notes[date];
      }
    },

    async createHabit(form: HabitForm) {
      await apiFetch<Habit>("/habits", { method: "POST", body: form });
      await this.reload();
    },

    async updateHabit(habitId: number, form: Partial<HabitForm> & { clear_reminder?: boolean }) {
      await apiFetch<Habit>(`/habits/${habitId}`, { method: "PATCH", body: form });
      await this.reload();
    },

    async deleteHabit(habitId: number) {
      await apiFetch(`/habits/${habitId}`, { method: "DELETE" });
      this.items = this.items.filter((i) => i.habit.id !== habitId);
    },

    async setArchived(habitId: number, archived: boolean) {
      await apiFetch(`/habits/${habitId}/${archived ? "archive" : "unarchive"}`, { method: "POST" });
      await this.reload();
    },

    async reorder(orderedIds: number[]) {
      const order = new Map(orderedIds.map((id, index) => [id, index]));
      this.items.sort((a, b) => (order.get(a.habit.id) ?? 0) - (order.get(b.habit.id) ?? 0));
      await apiFetch("/habits/positions", { method: "PUT", body: { ordered_ids: orderedIds } });
    },
  },
});

File: backend/app/api/routers/habits.py
from datetime import date
from typing import Literal

from fastapi import APIRouter, Query

from app.api.deps import CurrentUserDep, SessionDep
from app.api.schemas.habits import (
    EntryChangeOut,
    EntryOut,
    EntryUpsert,
    HabitCreate,
    HabitOut,
    HabitOverviewOut,
    HabitPatch,
    ReorderRequest,
)
from app.application.use_cases import entries as entries_uc
from app.application.use_cases import habits as habits_uc
from app.infrastructure.repositories.habit_repo import HabitRepo

router = APIRouter(prefix="/habits", tags=["habits"])


@router.get("", response_model=list[HabitOut])
async def list_habits(user: CurrentUserDep, session: SessionDep) -> list[HabitOut]:
    return await HabitRepo(session).list_for_user(user.id)


@router.get("/overview", response_model=list[HabitOverviewOut])
async def overview(
    user: CurrentUserDep,
    session: SessionDep,
    from_date: date = Query(alias="from"),
    to_date: date = Query(alias="to"),
    sort: Literal["manual", "name", "color", "score", "status"] = "manual",
) -> list[HabitOverviewOut]:
    items = await habits_uc.get_overview(session, user.id, from_date, to_date, sort)
    return [
        HabitOverviewOut(
            habit=HabitOut.model_validate(i.habit),
            score=i.score,
            streak=i.streak,
            entries=i.entries,
            notes=i.notes,
        )
        for i in items
    ]


@router.post("", response_model=HabitOut, status_code=201)
async def create_habit(body: HabitCreate, user: CurrentUserDep, session: SessionDep) -> HabitOut:
    return await habits_uc.create_habit(session, user.id, body.model_dump())


@router.get("/{habit_id}", response_model=HabitOut)
async def get_habit(habit_id: int, user: CurrentUserDep, session: SessionDep) -> HabitOut:
    return await HabitRepo(session).get_owned(habit_id, user.id)


@router.patch("/{habit_id}", response_model=HabitOut)
async def update_habit(
    habit_id: int, body: HabitPatch, user: CurrentUserDep, session: SessionDep
) -> HabitOut:
    fields = body.model_dump(exclude_none=True, exclude={"clear_reminder"})
    if body.clear_reminder:
        fields["reminder_hour"] = None
        fields["reminder_min"] = None
    return await habits_uc.update_habit(session, user.id, habit_id, fields)


@router.delete("/{habit_id}", status_code=204)
async def delete_habit(habit_id: int, user: CurrentUserDep, session: SessionDep) -> None:
    await habits_uc.delete_habit(session, user.id, habit_id)


@router.put("/positions", status_code=204)
async def reorder(body: ReorderRequest, user: CurrentUserDep, session: SessionDep) -> None:
    await habits_uc.reorder_habits(session, user.id, body.ordered_ids)


@router.get("/{habit_id}/entries", response_model=list[EntryOut])
async def get_entries(
    habit_id: int,
    user: CurrentUserDep,
    session: SessionDep,
    from_date: date = Query(alias="from"),
    to_date: date = Query(alias="to"),
) -> list[EntryOut]:
    computed = await entries_uc.get_computed_entries(session, user.id, habit_id, from_date, to_date)
    return [
        EntryOut(date=d, value=value, notes=notes)
        for d, (value, notes) in sorted(computed.items(), reverse=True)
    ]


@router.put("/{habit_id}/entries/{entry_date}", response_model=EntryChangeOut)
async def upsert_entry(
    habit_id: int, entry_date: date, body: EntryUpsert, user: CurrentUserDep, session: SessionDep
) -> EntryChangeOut:
    change = await entries_uc.upsert_entry(session, user.id, habit_id, entry_date, body.value, body.notes)
    return EntryChangeOut(**change.__dict__)


@router.post("/{habit_id}/entries/{entry_date}/toggle", response_model=EntryChangeOut)
async def toggle_entry(
    habit_id: int, entry_date: date, user: CurrentUserDep, session: SessionDep
) -> EntryChangeOut:
    change = await entries_uc.toggle_entry(session, user.id, habit_id, entry_date)
    return EntryChangeOut(**change.__dict__)


@router.delete("/{habit_id}/entries/{entry_date}", response_model=EntryChangeOut)
async def delete_entry(
    habit_id: int, entry_date: date, user: CurrentUserDep, session: SessionDep
) -> EntryChangeOut:
    change = await entries_uc.delete_entry(session, user.id, habit_id, entry_date)
    return EntryChangeOut(**change.__dict__)
